controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,215 @@
+// Pipeline.
+func (r *PipelineRunReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var pipelineRun tknv1beta1.PipelineRun
+	if err := r.Get(ctx, req.NamespacedName, &pipelineRun); err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Unable to fetch PipelineRun")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// creating a objectRef based on the informed PipelineRun, the instance is informed to the
+	// inventory query interface to list Shipwright Builds that should be triggered
+	objectRef, err := filter.PipelineRunToObjectRef(ctx, r.Clock.Now(), &pipelineRun)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	logger.V(0).Info(
+		"Searching for Builds matching criteria",
+		"ref-name", objectRef.Name,
+		"ref-status", objectRef.Status,
+		"ref-selector", objectRef.Selector,
+	)
+
+	// search for Builds with Pipeline triggers matching current ObjectRef criteria
+	buildsToBeIssued := r.buildInventory.SearchForObjectRef(v1alpha1.PipelineTrigger, objectRef)
+	if len(buildsToBeIssued) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	buildNames := inventory.ExtractBuildNames(buildsToBeIssued...)
+	logger.V(0).Info("Build names in the Inventory matching criteria", "build-names", buildNames)
+
+	// during pipeline re-run a new PipelineRun is issued based on a existing object copying over all
+	// the elements, including annotations. To allow re-runs we annotate the current object's name
+	// and only check previously triggered builds when the name matches
+	var triggeredBuilds = []filter.TriggeredBuild{}
+	if filter.PipelineRunAnnotatedNameMatchesObject(&pipelineRun) {
+		// extracting existing triggered-builds from the annotation, information needed to detect if
+		// the BuildRuns have already beeing issued for the PipelineRun
+		triggeredBuilds, err = filter.PipelineRunExtractTriggeredBuildsSlice(&pipelineRun)
+		if err != nil {
+			logger.V(0).Error(err, "parsing triggered-builds annotation")
+			// in case of errors an empty slice takes place, may incur the side effect of issuing
+			// duplicated BuildRuns
+			triggeredBuilds = []filter.TriggeredBuild{}
+		}
+
+		// filtering out the instances that have already been processed, the annotation extracted
+		// shows  which build names and the objectRef employed
+		if filter.TriggereBuildsContainsObjectRef(triggeredBuilds, buildNames, objectRef) {
+			logger.V(0).Info("BuildRuns for PipelineRun have already been issued!")
+			return ctrl.Result{}, nil
+		}
+	} else {
+		logger.V(0).Info("PipelineRun annotated name does not match current object!")
+	}
+	logger.V(0).Info("PipelineRun previously triggered builds", "triggered-builds", triggeredBuilds)
+
+	// firing the BuildRun instances for the informed Builds
+	buildRunsIssued, err := r.issueBuildRunsForPipelineRun(ctx, &pipelineRun, buildNames)
```

Comment:
You can here get an error together with a non-empty buildRunsIssued array. The current code will requeue the reconcile and that imo triggers the same Build again.

Reply:
We have a number of integration test cases making sure repeated `BuildRun` aren't issued by accident, so before reaching this part all the filtering took place. Could you try to test this logic and see if you can bypass the filtering? Please consider the integration tests.

Reply:
Indeed, however that's and unlikely event the Kubernetes API-Sever errors out in a single object. Currently, we are logging the `BuildRun` instances issued when this error occur, so we can assess later on.

In short, the controller can only react to the current context, and when the API-Server errors out I think the controller should retry.

WDYT?

Reply:
I think this will become relevant once we have a webhook that is taking over validations from the build controller and potentially rejects an object creation.

Imagine three builds with the same trigger where the first one gets successfully created but the second one fails on creation due to for example a missing parameter value.

We should then "commit" what was created successfully before returning the error for a retry.

For the time being, I agree it is requiring unexpected errors like API server blibs or similar.